src/checkpointTrie.js

```diff
@@ -111,60 +118,50 @@ module.exports = class CheckpointTrie extends BaseTrie {
   createScratchReadStream (scratch) {
     const trie = this.copy()
     scratch = scratch || this._scratch
+    scratch = new DB(scratch._db)
     // Only read from the scratch
-    trie._getDBs = [scratch]
+    trie.db = scratch
     trie._scratch = scratch
     return new ScratchReadStream(trie)
   }
 
+  putRaw (key, val, cb) {
+    if (this.isCheckpoint) return this._overridePutRaw(key, val, cb)
+    this.db.put(key, val, cb)
+  }
+
   /**
    * Puts kv-pair directly to db, ignoring checkpoints.
    * @private
    */
   _overridePutRaw (key, val, cb) {
-    const dbPut = (db, cb2) => {
-      db.put(key, val, {
-        keyEncoding: 'binary',
-        valueEncoding: 'binary'
-      }, cb2)
-    }
-    async.each(this.__putDBs, dbPut, cb)
+    this._mainDB.put(key, val, cb)
   }
 
   /**
    * Enter into checkpoint mode.
    * @private
    */
   _enterCpMode () {
-    this._scratch = level()
-    this._getDBs = [this._scratch].concat(this._getDBs)
-    this.__putDBs = this._putDBs
-    this._putDBs = [this._scratch]
-    this._putRaw = this.putRaw
-    this.putRaw = this._overridePutRaw
+    this._scratch = new ScratchDB(this._mainDB)
+    this.db = this._scratch
```

Comment:
We have some naming confusion between `DB` the class and `DB` the level instance, respectively this would be more easy to grasp if we would differentiate here.

Have no brilliant idea for a naming scheme here (prefixing all `level` stuff with `level` might also not be the best idea since eventually we want to support further types of storage in the future).

One idea: prefix with `Trie` for the class versions, since these have some trie related wrapping changes, so e.g. `TrieDB.js` and `this.trieDB`. Hmm, but doesn't like this yet too much myself though.

Any thoughts?

Follow-up thought: maybe this would be good to be named such as `DB` the class remains as simple as currently be since this is the main exposed part of the API.

Reply:
Yeah, that would make sense.